File: signature.js
//Require package crypto.js
const HmacSHA256 = require("crypto-js/hmac-sha256");
const SHA256 = require("crypto-js/sha256");
const Base64 = require("crypto-js/enc-base64");

/**
 * Sign request using HMAC-SHA256 schema.
 * @param {String} method GET, PUT, POST, DELETE
 * @param {String} path  path+query
 * @param {String} secret secret key
 * @param {String} timestamp http timestamp from Date or x-yoov-date header
 * @param {String} body Request content
 * @returns HMAC-SHA256 hashed Signature (base64 encoded)
 */
function sign(method, path, secret, timestamp, body) {
  //console.log(method, path, secret, timestamp, body);
  const verb = method.toUpperCase();
  const contentHash = SHA256(body).toString(Base64);
  //console.log("contentHash", contentHash);
  let strToBeSigned = `${verb}\n${path}\n${timestamp}\n${contentHash}`;
  return HmacSHA256(strToBeSigned, secret).toString(Base64);
}

File: signature.py
import base64
import hashlib
import hmac
import six

def sign(method, path, secret, timestamp, body):
    verb = method.upper()
    if six.PY2:
        content_digest = hashlib.sha256(bytes(body)).digest()
    else:
        content_digest = hashlib.sha256(bytes(body, 'utf-8')).digest()

    content_hash = base64.b64encode(content_digest).decode()

    string_to_sign = verb + '\n' + path + '\n' + \
        timestamp + '\n '+ content_hash

    if six.PY2:
        digest = hmac.new(bytes(secret), bytes(
            string_to_sign), hashlib.sha256).digest()
    else:
        digest = hmac.new(bytes(secret, 'utf-8'), bytes(
            string_to_sign, 'utf-8'), hashlib.sha256).digest()

    signature = base64.b64encode(digest).decode()

    return signature
